### PR TITLE
Added builder implementations to the required thermal input models

### LINKED ISSUE #165: Add copy builder implementations for thermal instances
Currently, copy implementations are not implemented for the following `AssetInput` entities and hence throw a `NotImplementedException`. It should be considered implementing the corresponding methods for these entities soon.

- CylindricalStorageInput
- ThermalBusInput
- ThermalHouseInput

### PR DESCRIPTION
**Pull request by ahamshubham**
Resolves #165. Changes done -

1. Added `CylindricalStorageInputCopyBuilder`, `ThermalBusInputCopyBuilder` and `ThermalHouseInputCopyBuilder` to classes `CylindricalStorageInput`, `ThermalBusInput` and `ThermalHouseInput` respectively.

Next, I will work on test cases and code refinement.

### CODE REVIEW

**Comment by ahamshubham** [comment]
Added test cases to check the builder implementations added to the thermal input models.

**Comment by ahamshubham** [comment]
!test

**Comment by ahamshubham** [comment]
!test

File: src/main/java/edu/ie3/datamodel/models/input/thermal/CylindricalStorageInput.java [hunk 1]
```diff
@@ -157,4 +155,76 @@ public String toString() {
...
+    }
+
+    @Override
+    public CylindricalStorageInput build() {
+      return new CylindricalStorageInput(
+          getUuid(),
+          getId(),
+          getOperator(),
+          getOperationTime(),
+          getThermalBus(),
```
Comment (ckittl) [inline src/main/java/edu/ie3/datamodel/models/input/thermal/CylindricalStorageInput.java:189]: You are pointing to the wrong thermal bus, here.

As your inner class is not static, you are able to point to the `CylindricalStorageInput#getThermalBus`, so more or less to the instance you wanna copy. Meaning, at the moment you are not able to alter the bus within the copy builder. You would need to point to a thermal bus within the copy builder-hierarchy.

As the thermal bus is used by several other classes as well, I would suggest to implement a new *abstract* class `ThermalUnitInpuiutCopyBuilder`  extending `AssetInputCopyBuilder` as inner class of `ThermalUnitInput`. It only holds a thermal bus as attribute and adds equivalent methods to set and get it in the same style as `AssetInputCopyBuilder`.

Then `CylindricalStorageInputCopyBuilder` and `ThermalHouseInputCopyBuilder` can extend the `ThermalUnitInputCopyBuilder`.

File: src/main/java/edu/ie3/datamodel/models/input/thermal/ThermalHouseInput.java [hunk 2]
```diff
@@ -109,4 +107,50 @@ public String toString() {
...
+    }
+
+    @Override
+    public ThermalHouseInput build() {
+      return new ThermalHouseInput(
+          getUuid(),
+          getId(),
+          getOperator(),
+          getOperationTime(),
+          getThermalBus(),
```
Comment (ckittl) [inline src/main/java/edu/ie3/datamodel/models/input/thermal/ThermalHouseInput.java:135]: Same as above.

File: src/test/groovy/edu/ie3/datamodel/models/input/thermal/CylindricalStorageInputTest.groovy [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+class CylindricalStorageInputTest extends Specification {
+
+	def "A CylindricalStorageInput copy method should work as expected"() {
+		given:
+		def cylindricalStorageInput = ThermalUnitInputTestData.cylindricStorageInput
+
+		when:
+		def alteredUnit = cylindricalStorageInput.copy().storageVolumeLvl(ThermalUnitInputTestData.storageVolumeLvl)
+				.storageVolumeLvlMin(ThermalUnitInputTestData.storageVolumeLvlMin).inletTemp(ThermalUnitInputTestData.inletTemp)
+				.returnTemp(ThermalUnitInputTestData.returnTemp).c(ThermalUnitInputTestData.c).build()
```
Comment (ckittl) [inline src/test/groovy/edu/ie3/datamodel/models/input/thermal/CylindricalStorageInputTest.groovy:21]: Test case for altered thermal bus is missing.

File: src/test/groovy/edu/ie3/datamodel/models/input/thermal/ThermalHouseInputTest.groovy [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+class ThermalHouseInputTest extends Specification {
+
+	def "A ThermalHouseInput copy method should work as expected"() {
+		given:
+		def thermalHouseInput = ThermalUnitInputTestData.thermalHouseInput
+
+		when:
+		def alteredUnit = thermalHouseInput.copy().ethLosses(ThermalUnitInputTestData.thermalConductance)
+				.ethCapa(ThermalUnitInputTestData.ethCapa).build()
```
Comment (ckittl) [inline src/test/groovy/edu/ie3/datamodel/models/input/thermal/ThermalHouseInputTest.groovy:20]: Test case for thermal bus is missing.

**Review by ckittl — CHANGES_REQUESTED** [review]

File: src/main/java/edu/ie3/datamodel/models/input/thermal/CylindricalStorageInput.java [hunk 5]
```diff
@@ -157,4 +155,76 @@ public String toString() {
         + c
         + '}';
   }
+
+  /**
+   * A builder pattern based approach to create copies of {@link CylindricalStorageInput} entities
+   * with altered field values. For detailed field descriptions refer to java docs of {@link
+   * CylindricalStorageInput}
+   */
+  public class CylindricalStorageInputCopyBuilder
```
[the remark this answers is not in the document]
Comment (ahamshubham) [reply src/main/java/edu/ie3/datamodel/models/input/thermal/CylindricalStorageInput.java:164]: Will make the change.

File: src/main/java/edu/ie3/datamodel/models/input/thermal/ThermalBusInput.java [hunk 6]
```diff
@@ -36,9 +35,30 @@ public ThermalBusInput(UUID uuid, String id) {
...
+  public ThermalBusInputCopyBuilder copy() {
+    return new ThermalBusInputCopyBuilder(this);
+  }
+
+  /**
+   * A builder pattern based approach to create copies of {@link ThermalBusInput} entities with
+   * altered field values. For detailed field descriptions refer to java docs of {@link
+   * ThermalBusInput}
+   */
+  public class ThermalBusInputCopyBuilder
```
[the remark this answers is not in the document]
Comment (ahamshubham) [reply src/main/java/edu/ie3/datamodel/models/input/thermal/ThermalBusInput.java:47]: Will make the change.

File: src/main/java/edu/ie3/datamodel/models/input/thermal/ThermalHouseInput.java [hunk 7]
```diff
@@ -109,4 +107,50 @@ public String toString() {
         + ethCapa
         + '}';
   }
+
+  /**
+   * A builder pattern based approach to create copies of {@link ThermalHouseInput} entities with
+   * altered field values. For detailed field descriptions refer to java docs of {@link
+   * ThermalHouseInput}
+   */
+  public class ThermalHouseInputCopyBuilder
```
[the remark this answers is not in the document]
Comment (ahamshubham) [reply src/main/java/edu/ie3/datamodel/models/input/thermal/ThermalHouseInput.java:116]: Will make the change.

File: src/main/java/edu/ie3/datamodel/models/input/thermal/CylindricalStorageInput.java (hunk 1 above)
Comment (ahamshubham) [reply src/main/java/edu/ie3/datamodel/models/input/thermal/CylindricalStorageInput.java:189]: Sure, will make the changes.

File: src/main/java/edu/ie3/datamodel/models/input/thermal/ThermalHouseInput.java (hunk 2 above)
Comment (ahamshubham) [reply src/main/java/edu/ie3/datamodel/models/input/thermal/ThermalHouseInput.java:135]: Sure, will make the changes.

File: src/test/groovy/edu/ie3/datamodel/models/input/thermal/CylindricalStorageInputTest.groovy (hunk 3 above)
Comment (ahamshubham) [reply src/test/groovy/edu/ie3/datamodel/models/input/thermal/CylindricalStorageInputTest.groovy]: Added.

File: src/test/groovy/edu/ie3/datamodel/models/input/thermal/ThermalHouseInputTest.groovy (hunk 4 above)
Comment (ahamshubham) [reply src/test/groovy/edu/ie3/datamodel/models/input/thermal/ThermalHouseInputTest.groovy]: Added.

**Review by ckittl — APPROVED** [review]

**Pull request merged**